### DISCUSSION TITLE
Porting to FreeBSD

### DISCUSSION THREAD

**Discussion by raphaelahrens** [Try/OpenGothic#413] [Show and tell]
Hi,

I wanted to share that I was able to port OpenGothic to FreeBSD.
After finding out that I had to use GNUmake and GCC and also modifing the CMakefile and some code, I was able to compile it and also play it. Already maded it to the upper city. The port works but is sometimes a little bit laggy, especially inside of buildings. By reducing the maximum number of threads it improved but is still there.

Would you be interested in the changes I had to make to compile it for FreeBSD?

**Comment by Try** [comment]
Hi, @raphaelahrens !

Yes, please submit PR. Portability is important for this project)

> By reducing the maximum number of threads it improved but is still there.

Hm, interesting... multi-threading is there primarily for animation, so should not interact with indoor/outdoor cases.

PS: since it's show-and-tell screenshoots are welcome ;)

**Reply by raphaelahrens** [reply]
Ok I created some PRs for both Tempest and OpenGothic

Regarding the performance, I finally found out how to show the framerate and it is at around 20 fps, so I guess this is part of the issue.
What is weird is that the threats are not at 100%, mhh could be a problem of my setup here.
Reducing the resolution helps with the fps, but there are still parts where the game stops and the the fps drop.
Do you have some settings I can try to improve the performance?

Here the requested  
![OpenGothic](https://user-images.githubusercontent.com/5632364/220358482-21e801a1-1943-45bd-9418-42b633b02d30.png)
screenshot

**Reply by Try** [reply]
> What is weird is that the threats are not at 100%, mhh could be a problem of my setup here.

This is probably fine - there is no a lot of work that can run in parallel. Only frustum culling and animation uses threading, while rest of application is single threaded. 

> Do you have some settings I can try to improve the performance?

You may want to check "Extended configuration" settings in menu:
"cloud shadows" is SSAO - quite expensive 
"radial fog" also worth checking 
Bot of this are quite intensive compute shaders.

Other than that.. well probably wont optimize for GT1030, since it's old Pascal architecture and mesh-shading is available only since Turing.

**Reply by raphaelahrens** [reply]
I was able to further pin down some of my problems.
With a lower resolution I was able to improve the frame rate, which is ok for me.

But the issue with the radical drop in the frame rate remained.
While the game is playable it is annoying when you are in a fight and suddenly you have a freeze frame.

So I created a flamegraph 
![test](https://user-images.githubusercontent.com/5632364/221442312-7c4d9e36-7f78-4812-a7d8-8c0174933393.png)


You can see that around 14% of the time is wasted on waiting for a XReply from the X11 server.
I further looked into what is going on there and found that the Engine is constantly checking if the window is in fullscreen mode and if the window has been resized.

Further I found out that in [mainwindow.cpp:273 ](https://github.com/Try/OpenGothic/blob/master/game/mainwindow.cpp#L273) everytime the mouse is moved it first checks if the window is in fullscreen.

The issue is that for X11 the Engine uses XGetWindow{Attribute,Properties} which both are quite expensive calls to the XServer for properties which should not change often while playing.

I removed some of the fullscreen checks and saw a drop in the CPU utilization of the Xorg process and a better performance of the game overall.

I'm not sure how to really fix this, maybe it makes sense to cache these values and only pull on an interval.

Anyway just wanted to share what I found.

**Reply by Try** [reply]
> Anyway just wanted to share what I found.

Thanks! Very cool finding, so let me break it into cases:
* Periodic `XGetWindow` - this is workaround to emulate resizeEvent, not quite remember what was the issue with it on Linux
* `isFullScreen` - is builtin feature on Windows and Mac and practically zero-cost. Such performance hit is quite annoying here 

Currently checking, if there is a way to implement event-based handling of this on X11

**Reply by Try** [reply]
Pushed 05e7eec into Tempest - fix for resize event

**Reply by Try** [reply]
Should be fixed(377a51b2) in latest OpenGothic commit
